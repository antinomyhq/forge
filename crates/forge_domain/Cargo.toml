[package]
name = "forge_domain"
version = "0.1.0"
edition = "2021"

[dependencies]
handlebars = "6.0.0"
Inflector = "0.11.4"
anyhow = "1.0"
async-trait = "0.1.84"
chrono = { version = "0.4.39", features = ["serde"] }
derive_more = { version = "1.0.0", features = ["from", "display"] }
derive_setters = "0.1.6"
futures = "0.3.31"
html-escape = "0.2.13"
nom = "8.0.0"
quick-xml = { version = "0.37.2", features = ["serialize"] }
schemars = "0.8.21"
serde = "1.0.217"
serde_json = "1.0.134"
strum = "0.26.3"
strum_macros = "0.26.4"
thiserror = "2.0.11"
tokio = { version = "1.42.0", features = ["full", "test-util"] }
tokio-stream = "0.1.17"
uuid = { version = "1.11.0", features = [
    "v4",
    "fast-rng",
    "macro-diagnostics",
    "serde",
] }

[dev-dependencies]
insta = { version = "1.34.0", features = ["json"] }
pretty_assertions = "1.4.1"
