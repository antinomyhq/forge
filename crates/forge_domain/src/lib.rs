mod agent;
mod arena;
mod chat_request;
mod chat_response;
mod chat_stream_ext;
mod command;
mod config;
mod context;
mod context_ext;
mod conversation;
mod env;
mod errata;
mod error;
mod file;
mod learning;
mod message;
mod model;
mod provider;
mod stream_ext;
mod tool;
mod tool_call;
mod tool_call_parser;
mod tool_choice;
mod tool_definition;
mod tool_name;
mod tool_result;
mod tool_service;
mod tool_usage;
mod user_interaction;
mod workflow;

pub use agent::*;
pub use chat_request::*;
pub use chat_response::*;
pub use chat_stream_ext::*;
pub use command::*;
pub use config::*;
pub use context::*;
pub use context_ext::*;
pub use conversation::*;
pub use env::*;
pub use errata::*;
pub use error::*;
pub use file::*;
pub use learning::*;
pub use message::*;
pub use model::*;
pub use provider::*;
pub use stream_ext::*;
pub use tool::*;
pub use tool_call::*;
pub use tool_call_parser::*;
pub use tool_choice::*;
pub use tool_definition::*;
pub use tool_name::*;
pub use tool_result::*;
pub use tool_service::*;
pub use tool_usage::*;
pub use user_interaction::*;
pub use workflow::*;

/// Core domain trait providing access to services and repositories.
/// This trait follows clean architecture principles for dependency management
/// and service/repository composition.
pub trait ForgeDomain {
    /// The concrete type implementing file read service capabilities
    type FileReadService: file::FileReadService;
    /// The concrete type implementing tool service capabilities
    type ToolService: ToolService;
    /// The concrete type implementing provider service capabilities
    type ProviderService: ProviderService;
    /// The concrete type implementing conversation repository
    type ConversationRepo: ConversationRepository;
    /// The concrete type implementing configuration repository
    type ConfigRepo: ConfigRepository;
    /// The concrete type implementing environment repository
    type EnvironmentRepo: EnvironmentRepository;

    /// Get a reference to the tool service instance
    fn tool_service(&self) -> &Self::ToolService;
    /// Get a reference to the provider service instance
    fn provider_service(&self) -> &Self::ProviderService;
    /// Get a reference to the conversation repository instance
    fn conversation_repository(&self) -> &Self::ConversationRepo;
    /// Get a reference to the configuration repository instance
    fn config_repository(&self) -> &Self::ConfigRepo;
    /// Get a reference to the environment repository instance
    fn environment_repository(&self) -> &Self::EnvironmentRepo;

    /// Get a reference to the file read service instance
    fn file_read_service(&self) -> &Self::FileReadService;
}
