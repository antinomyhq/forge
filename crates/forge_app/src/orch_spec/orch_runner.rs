use std::collections::{HashMap, VecDeque};
use std::path::PathBuf;
use std::sync::Arc;

use chrono::Local;
use derive_setters::Setters;
use forge_domain::{
    Agent, AgentId, ChatCompletionMessage, ChatResponse, ContextMessage, Conversation,
    ConversationId, Environment, Event, HttpConfig, ModelId, RetryConfig, Role, Template,
    ToolCallFull, ToolResult, Workflow,
};
use handlebars::{Handlebars, no_escape};
use rust_embed::Embed;
use tokio::sync::Mutex;
use tokio::sync::mpsc::Sender;
use url::Url;

use crate::AgentService;
use crate::orch::Orchestrator;

#[derive(Embed)]
#[folder = "../../templates/"]
struct Templates;

struct Runner {
    hb: Handlebars<'static>,
    // History of all the updates made to the conversation
    conversation_history: Mutex<Vec<Conversation>>,

    // Tool call requests and the mock responses
    test_tool_calls: Mutex<VecDeque<(ToolCallFull, ToolResult)>>,

    // Mock completions from the LLM (Each value is produced as an event in the stream)
    test_completions: Mutex<VecDeque<ChatCompletionMessage>>,
}

impl Runner {
    fn new(setup: &Setup) -> Self {
        let mut hb = Handlebars::new();
        hb.set_strict_mode(true);
        hb.register_escape_fn(no_escape);

        // Register all partial templates
        hb.register_embed_templates::<Templates>().unwrap();
        for (name, tpl) in &setup.templates {
            hb.register_template_string(name, tpl).unwrap();
        }

        Self {
            hb,
            conversation_history: Mutex::new(Vec::new()),
            test_tool_calls: Mutex::new(VecDeque::from(setup.mock_tool_call_responses.clone())),
            test_completions: Mutex::new(VecDeque::from(setup.mock_assistant_responses.clone())),
        }
    }

    // Returns the conversation history
    async fn get_history(&self) -> Vec<Conversation> {
        self.conversation_history.lock().await.clone()
    }
}

#[async_trait::async_trait]
impl AgentService for Runner {
    async fn chat_agent(
        &self,
        _id: &forge_domain::ModelId,
        _context: forge_domain::Context,
    ) -> forge_domain::ResultStream<ChatCompletionMessage, anyhow::Error> {
        let mut responses = self.test_completions.lock().await;
        if let Some(message) = responses.pop_front() {
            Ok(Box::pin(tokio_stream::iter(std::iter::once(Ok(message)))))
        } else {
            Ok(Box::pin(tokio_stream::iter(std::iter::empty())))
        }
    }

    async fn call(
        &self,
        _agent: &forge_domain::Agent,
        _context: &mut forge_domain::ToolCallContext,
        test_call: forge_domain::ToolCallFull,
    ) -> forge_domain::ToolResult {
        let mut guard = self.test_tool_calls.lock().await;
        for (id, (call, result)) in guard.iter().enumerate() {
            if call.call_id == test_call.call_id {
                let result = result.clone();
                guard.remove(id);
                return result;
            }
        }
        panic!("Tool call not found")
    }

    async fn render(
        &self,
        template: &str,
        object: &(impl serde::Serialize + Sync),
    ) -> anyhow::Result<String> {
        Ok(self.hb.render_template(template, object)?)
    }

    async fn update(&self, conversation: Conversation) -> anyhow::Result<()> {
        self.conversation_history.lock().await.push(conversation);
        Ok(())
    }
}

fn new_orchestrator(
    setup: &Setup,
    tx: Sender<anyhow::Result<ChatResponse>>,
) -> (Orchestrator<Runner>, Arc<Runner>) {
    let services = Arc::new(Runner::new(setup));
    let environment = new_env();
    let conversation = Conversation::new(
        ConversationId::generate(),
        setup.workflow.clone(),
        Default::default(),
    );
    let current_time = Local::now();

    let orch = Orchestrator::new(services.clone(), environment, conversation, current_time)
        .sender(Arc::new(tx))
        .files(setup.files.clone());

    // Return setup
    (orch, services)
}

fn new_env() -> Environment {
    Environment {
        os: "MacOS".to_string(),
        pid: 1234,
        cwd: PathBuf::from("/Users/tushar"),
        home: Some(PathBuf::from("/Users/tushar")),
        shell: "bash".to_string(),
        base_path: PathBuf::from("/Users/tushar/projects"),
        forge_api_url: Url::parse("http://localhost:8000").unwrap(),
        retry_config: RetryConfig::default(),
        max_search_lines: 1000,
        fetch_truncation_limit: 1024,
        stdout_max_prefix_length: 256,
        stdout_max_suffix_length: 256,
        max_read_size: 4096,
        http: HttpConfig::default(),
        max_file_size: 1024 * 1024 * 5,
        max_search_result_bytes: 200,
        stdout_max_line_length: 200, // 5 MB
    }
}

async fn run(setup: Setup) -> TestContext {
    const LIMIT: usize = 1024;
    let mut chat_responses = Vec::new();
    let (tx, mut rx) = tokio::sync::mpsc::channel::<anyhow::Result<ChatResponse>>(LIMIT);
    let (mut orch, runner) = new_orchestrator(&setup, tx);

    tokio::join!(
        async { orch.chat(setup.event).await.unwrap() },
        rx.recv_many(&mut chat_responses, LIMIT)
    );
    TestContext {
        conversation_history: runner.get_history().await,
        chat_responses,
    }
}

// The final output produced after running the orchestrator to completion
#[derive(Debug)]
pub struct TestContext {
    pub conversation_history: Vec<Conversation>,
    pub chat_responses: Vec<anyhow::Result<ChatResponse>>,
}

impl TestContext {
    pub fn system_prompt(&self) -> Option<&str> {
        self.conversation_history
            .last()
            .and_then(|c| c.context.as_ref())
            .and_then(|c| c.messages.iter().find(|c| c.has_role(Role::System)))
            .and_then(|c| c.content())
    }

    pub fn context_messages(&self) -> Vec<ContextMessage> {
        self.conversation_history
            .last()
            .and_then(|c| c.context.as_ref())
            .map(|c| c.messages.clone())
            .clone()
            .unwrap_or_default()
    }
}

#[derive(Setters)]
#[setters(into)]
pub struct Setup {
    pub event: Event,
    pub mock_tool_call_responses: Vec<(ToolCallFull, ToolResult)>,
    pub mock_assistant_responses: Vec<ChatCompletionMessage>,
    pub workflow: Workflow,
    pub templates: HashMap<String, String>,
    pub files: Vec<String>,
}

impl Setup {
    pub fn init_forge_task(task: &str) -> Self {
        Self::from_event(Event::new("forge/user_task_init", Some(task)))
    }

    pub fn from_event(event: Event) -> Self {
        Self {
            event,
            mock_assistant_responses: Default::default(),
            mock_tool_call_responses: Default::default(),
            workflow: Workflow::new()
                .model(ModelId::new("openai/gpt-1"))
                .agents(vec![
                    Agent::new(AgentId::new("forge"))
                        .system_prompt(Template::new("You are Forge"))
                        .tools(vec![("fs_read").into(), ("fs_write").into()]),
                    Agent::new(AgentId::new("must"))
                        .system_prompt(Template::new("You are Muse"))
                        .tools(vec![("fs_read").into()]),
                ])
                .tool_supported(true),
            templates: Default::default(),
            files: Default::default(),
        }
    }

    pub async fn run(self) -> TestContext {
        run(self).await
    }
}
