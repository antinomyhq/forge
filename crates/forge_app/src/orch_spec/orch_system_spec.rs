use forge_domain::{ChatCompletionMessage, Content, Workflow};
use insta::assert_snapshot;

use crate::orch_spec::orch_runner::TestContext;

#[tokio::test]
async fn test_system_prompt() {
    let mut ctx = TestContext::default()
        .workflow(Workflow::default().tool_supported(false))
        .mock_assistant_responses(vec![ChatCompletionMessage::assistant(Content::full(
            "Sure",
        ))]);

    ctx.run("This is a test").await.unwrap();
    let system_messages = ctx.output.system_messages().unwrap().join("\n\n");
    assert_snapshot!(system_messages);
}

#[tokio::test]
async fn test_system_prompt_tool_supported() {
    let mut ctx = TestContext::default()
        .workflow(
            Workflow::default()
                .tool_supported(true)
                .custom_rules("Do it nicely"),
        )
        .files(vec![
            "/users/john/foo.txt".to_string(),
            "/users/jason/bar.txt".to_string(),
        ])
        .mock_assistant_responses(vec![ChatCompletionMessage::assistant(Content::full(
            "Sure",
        ))]);

    ctx.run("This is a test").await.unwrap();

    let system_messages = ctx.output.system_messages().unwrap().join("\n\n");
    assert_snapshot!(system_messages);
}
