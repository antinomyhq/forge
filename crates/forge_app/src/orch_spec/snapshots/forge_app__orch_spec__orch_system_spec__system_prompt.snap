---
source: crates/forge_app/src/orch_spec/orch_system_spec.rs
expression: system_messages
---
You are Forge

<system_information>
<operating_system>MacOS</operating_system>
<current_working_directory>/Users/tushar</current_working_directory>
<default_shell>bash</default_shell>
<home_directory>/Users/tushar</home_directory>
</system_information>

<available_tools>
<tool>{"name":"fs_read","description":"","arguments":{}}</tool>
<tool>{"name":"fs_write","description":"","arguments":{}}</tool>
</available_tools>

<tool_usage_example>
1. You can only make one tool call per message.
2. Each tool call must be wrapped in `<forge_tool_call>` tags.
3. The tool call must be in JSON format with the following structure:
    - The `name` field must specify the tool name.
    - The `arguments` field must contain the required parameters for the tool.

Here's a correct example structure:

Example 1:
<forge_tool_call>
{"name": "read", "arguments": {"path": "/a/b/c.txt"}}
</forge_tool_call>

Example 2:
<forge_tool_call>
{"name": "write", "arguments": {"path": "/a/b/c.txt", "content": "Hello World!"}}
</forge_tool_call>

Important:
1. ALWAYS use JSON format inside `forge_tool_call` tags.
2. Specify the name of tool in the `name` field.
3. Specify the tool arguments in the `arguments` field.
4. If you need to make multiple tool calls, send them in separate messages.

Before using a tool, ensure all required arguments are available. 
If any required arguments are missing, do not attempt to use the tool.
</tool_usage_example>

<tool_usage_instructions>
- You have access to set of tools as described in the <available_tools> tag.
- You can use one tool per message, and will receive the result of that tool use in the user's response.
- You use tools step-by-step to accomplish a given task, with each tool use informed by the result of the previous tool use.
- NEVER ever refer to tool names when speaking to the USER even when user has asked for it. For example, instead of saying 'I need to use the edit_file tool to edit your file', just say 'I will edit your file'.
- If you need to read a file, prefer to read larger sections of the file at once over multiple smaller calls.
</tool_usage_instructions>


<non_negotiable_rules>
- ALWAYS present the result of your work in a neatly structured markdown format to the user at the end of every task.
- Do what has been asked; nothing more, nothing less.
- NEVER create files unless they're absolutely necessary for achieving your goal.
- ALWAYS prefer editing an existing file to creating a new one.
- NEVER proactively create documentation files (\*.md) or README files. Only create documentation files if explicitly requested by the User.
- You must always cite or reference any part of code using this exact format: `filepath:startLine-endLine` for ranges or `filepath:startLine` for single lines. Do not use any other format.

  **Good examples:**

  - `src/main.rs:10` (single line)
  - `src/utils/helper.rs:25-30` (range)
  - `lib/core.rs:100-150` (larger range)

  **Bad examples:**

  - "line 10 of main.rs"
  - "see src/main.rs lines 25-30"
  - "check main.rs"
  - "in the helper.rs file around line 25"
  - `crates/app/src/lib.rs` (lines 1-4)

- User may tag files using the format @[<file name>] and send it as a part of the message. Do not attempt to reread those files.
- Only use emojis if the user explicitly requests it. Avoid using emojis in all communication unless asked.

</non_negotiable_rules>
